File: 0014-longest-common-prefix/0014-longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        def merge(word1, word2):
            m = min(len(word1), len(word2))
            i = 0
            while i < m:
                if word1[i] != word2[i]:
                    break
                i += 1
            return word1[:i]

        def divide(left, right):
            if left == right:
                return strs[left]
            mid = left + (right - left) // 2
            left_prefix = divide(left, mid)
            right_prefix = divide(mid + 1, right)
            return merge(left_prefix, right_prefix)

        return divide(0, len(strs) - 1)

File: 0014-longest-common-prefix/0014-longest-common-prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        return divide(strs, 0, strs.size() - 1);
    }

    string divide(vector<string>& strs, int l, int r) {
        if (l == r)
            return strs[l];
        int m = l + (r - l) / 2;
        string left_prefix = divide(strs, l, m);
        string right_prefix = divide(strs, m + 1, r);
        return merge(left_prefix, right_prefix);
    }

    string merge(string l, string r) {
        int i = 0;
        while (i < min(l.size(), r.size()) && l[i] == r[i]) {
            i++;
        }
        return l.substr(0, i);
    }
};